File: test/test_clients.js
var Scenario = require('./scenario'),
    faye     = require('../build/faye-node'),
    assert   = require('assert');

Scenario.run("Client modifies incoming messages",
function() { with(this) {
  server(8000);
  httpClient('A', ['/channels/a']);
  httpClient('B', ['/channels/b']);
  
  extendClient('A', 'incoming', function(message, callback) {
    if (message.data) message.data.modified = 'hi';
    callback(message);
  });
  
  publish('B', '/channels/a', {welcome: 'message'});
  checkInbox({
      A: {
        '/channels/a': [{welcome: 'message', modified: 'hi'}]
      },
      B: {}
  });
}});

Scenario.run("Client blocks incoming messages",
function() { with(this) {
  server(8000);
  httpClient('A', ['/channels/a']);
  httpClient('B', ['/channels/b']);
  
  extendClient('A', 'incoming', function(message, callback) {
    callback(null);
  });
  
  publish('B', '/channels/a', {welcome: 'message'});
  checkInbox({ A: {}, B: {} });
}});

Scenario.run("Server requires authentication",
function() { with(this) {
  server(8000);
  httpClient('A', ['/channels/a']);
  httpClient('B', ['/channels/b']);
  
  extendServer('incoming', function(message, callback) {
    if (message.ext && message.ext.password) callback(message);
  });
  
  extendClient('B', 'outgoing', function(message, callback) {
    message.ext = {password: true};
    callback(message);
  });
  
  publish('A', '/channels/b', {messageFor: 'B'});
  checkInbox({ A: {}, B: {} });
  
  publish('B', '/channels/a', {messageFor: 'A'});
  checkInbox({
      A: {
        '/channels/a': [{messageFor: 'A'}]
      },
      B: {}
  });
}});

Scenario.run("Server blocks a message by setting an error",
function() { with(this) {
  server(8000);
  httpClient('A', ['/channels/a']);
  httpClient('B', ['/channels/b']);
  
  extendServer('incoming', function(message, callback) {
    if (message.data) message.error = Faye.Error.extMismatch();
    callback(message);
  });
  
  listenForErrors('A');
  
  publish('A', '/channels/b', {messageFor: 'B'});
  checkInbox({ A: {}, B: {} });
  checkErrors('A', ['302::Extension mismatch']);
}});

Scenario.run("Server modifies outgoing message",
function() { with(this) {
  server(8000);
  httpClient('A', []);
  httpClient('B', ['/channels/b']);
  
  extendServer('outgoing', function(message, callback) {
    if (message.data) message.data.addition = 56;
    callback(message);
  });
  
  publish('A', '/channels/b', {messageFor: 'B'});
  checkInbox({
      A: {},
      B: {
          '/channels/b': [{messageFor: 'B', addition: 56}]
      }
  });
}});

['outgoing', 'incoming'].forEach(function(direction) {
  Scenario.run("Server delays " + direction + " message",
  function() { with(this) {
    server(8000);
    httpClient('A', []);
    httpClient('B', ['/channels/b']);
    
    extendServer(direction, function(message, callback) {
      var timeout = message.data ? 5000 : 0;
      setTimeout(function() { callback(message) }, timeout);
    });
    
    publish('A', '/channels/b', {messageFor: 'B'});
    checkInbox({ A: {}, B: {} });
    
    wait(4);
    checkInbox({ A: {}, B: {} });
    
    wait(1);
    
    checkInbox({
        A: {},
        B: {
            '/channels/b': [{messageFor: 'B'}]
        }
    });
  }});
});

Scenario.run("Server blocks outgoing message",
function() { with(this) {
  server(8000);
  httpClient('A', []);
  httpClient('B', ['/channels/b']);
  
  extendServer('outgoing', function(message, callback) {
    if (!message.data) return callback(message);
    if (message.data.deliver === 'yes') return callback(message);
    callback(null);
  });
  
  publish('A', '/channels/b', [{deliver: 'no'}, {deliver: 'yes'}]);
  
  checkInbox({
      A: {},
      B: {
          '/channels/b': [{deliver: 'yes'}]
      }
  });
}});

Scenario.run("Server goes away, subscriptions should be revived",
function() { with(this) {
  server(8000);
  httpClient('A', ['/channels/a']);
  httpClient('B', []);
  killServer();
  wait(6);
  server(8000);
  wait(15);
  publish('B', '/channels/a', {hello: 'world'});
  checkInbox({
      A: {
        '/channels/a': [{hello: 'world'}]
      },
      B: {}
  });
}});

Scenario.run("Two HTTP clients, no messages delivered",
function() { with(this) {
  server(8000);
  httpClient('A', ['/channels/a']);
  httpClient('B', []);
  publish('B', '/channels/b', {hello: 'world'});
  checkInbox({
      A: {},
      B: {}
  });
}});

Scenario.run("Two HTTP clients, single subscription",
function() { with(this) {
  server(8000);
  httpClient('A', ['/channels/a']);
  httpClient('B', []);
  publish('B', '/channels/a', {hello: 'world'});
  checkInbox({
      A: {
        '/channels/a': [{hello: 'world'}]
      },
      B: {}
  });
}});

Scenario.run("Two HTTP clients, two identical messages sent together",
function() { with(this) {
  server(8000);
  httpClient('A', ['/channels/a']);
  httpClient('B', []);
  publish('B', '/channels/a', [{hello: 'world'}, {hello: 'world'}]);
  checkInbox({
      A: {
        '/channels/a': [{hello: 'world'}, {hello: 'world'}]
      },
      B: {}
  });
}});

Scenario.run("Two HTTP clients, two message deliveries",
function() { with(this) {
  server(8000);
  httpClient('A', ['/channels/a']);
  httpClient('B', []);
  publish('B', '/channels/a', {hello: 'world'});
  checkInbox({
      A: {
        '/channels/a': [{hello: 'world'}]
      },
      B: {}
  });
  wait(1);
  publish('B', '/channels/a', {hello: 'world'});
  checkInbox({
      A: {
        '/channels/a': [{hello: 'world'}, {hello: 'world'}]
      },
      B: {}
  });
}});

Scenario.run("Two HTTP clients, multiple subscriptions",
function() { with(this) {
  server(8000);
  httpClient('A', ['/channels/a', '/channels/*']);
  httpClient('B', []);
  publish('B', '/channels/a', {hello: 'world'});
  checkInbox({
      A: {
        '/channels/a': [{hello: 'world'}],
        '/channels/*': [{hello: 'world'}]
      },
      B: {}
  });
}});

Scenario.run("Two HTTP clients, two subscriptions on the same channel",
function() { with(this) {
  server(8000);
  httpClient('A', ['/channels/a']);
  httpClient('B', []);
  subscribe('A', '/channels/a');
  publish('B', '/channels/a', {hello: 'world'});
  checkInbox({
      A: {
        '/channels/a': [{hello: 'world'}, {hello: 'world'}]
      },
      B: {}
  });
}});

Scenario.run("Two HTTP clients, two subscriptions and one unsubscription",
function() { with(this) {
  server(8000);
  httpClient('A', ['/channels/a']);
  httpClient('B', []);
  subscribe('A', '/channels/a');
  cancelLastSubscription();
  publish('B', '/channels/a', {another: 'message'});
  checkInbox({
      A: {
        '/channels/a': [{another: 'message'}]
      },
      B: {}
  });
}});

Scenario.run("Three HTTP clients, single receiver",
function() { with(this) {
  server(8000);
  httpClient('A', ['/channels/a']);
  httpClient('B', []);
  httpClient('C', ['/channels/c']);
  publish('B', '/channels/a', {chunky: 'bacon'});
  checkInbox({
      A: {
        '/channels/a': [{chunky: 'bacon'}]
      },
      B: {},
      C: {}
  });
}});

Scenario.run("Three HTTP clients, multiple receivers",
function() { with(this) {
  server(8000);
  httpClient('A', ['/channels/shared']);
  httpClient('B', []);
  httpClient('C', ['/channels/shared']);
  publish('B', '/channels/shared', {chunky: 'bacon'});
  checkInbox({
      A: {
        '/channels/shared': [{chunky: 'bacon'}]
      },
      B: {},
      C: {
        '/channels/shared': [{chunky: 'bacon'}]
      }
  });
}});

Scenario.run("Two HTTP clients, single wildcard on receiver",
function() { with(this) {
  server(8000);
  httpClient('A', ['/channels/*', '/channels/anything']);
  httpClient('B', []);
  publish('B', '/channels/anything', {msg: 'hey'});
  checkInbox({
      A: {
        '/channels/*': [{msg: 'hey'}],
        '/channels/anything': [{msg: 'hey'}]
      },
      B: {}
  });
}});

File: test/test_clients.rb
require "test/unit"
require File.dirname(__FILE__) + "/../lib/faye"
require "test/scenario"

class TestClients < Test::Unit::TestCase
  include Faye
  include Scenario
  
  scenario "Client modifies incoming messages" do
    server 8000
    http_client :A, ['/channels/a']
    http_client :B, ['/channels/b']
    
    extend_client :A, :incoming, lambda { |message, callback|
      message['data']['modified'] = 'hi'
      callback.call(message)
    }
    
    publish :B, '/channels/a', 'welcome' => 'message'
    check_inbox(
        :A => {
          '/channels/a' => ['welcome' => 'message', 'modified' => 'hi']
        },
        :B => {}
    )
  end
  
  scenario "Client blocks incoming messages" do
    server 8000
    http_client :A, ['/channels/a']
    http_client :B, ['/channels/b']
    
    extend_client :A, :incoming, lambda { |message, callback|
      callback.call(nil)
    }
    
    publish :B, '/channels/a', 'welcome' => 'message'
    check_inbox( :A => {}, :B => {} )
  end
  
  scenario "Server requires authentication" do
    server 8000
    http_client :A, ['/channels/a']
    http_client :B, ['/channels/b']
    
    extend_server :incoming, lambda { |message, callback|
      if message['ext'] and message['ext']['password']
        callback.call(message)
      end
    }
    
    extend_client :B, :outgoing, lambda { |message, callback|
      message['ext'] = {'password' => true}
      callback.call(message)
    }
    
    publish :A, '/channels/b', 'message_for' => 'B'
    check_inbox( :A => {}, :B => {} )
    
    publish :B, '/channels/a', 'message_for' => 'A'
    check_inbox(
        :A => {
          '/channels/a' => ['message_for' => 'A']
        },
        :B => {}
    )
  end
  
  scenario "Server modifies outgoing message" do
    server 8000
    http_client :A, []
    http_client :B, ['/channels/b']
    
    extend_server :outgoing, lambda { |message, callback|
      message['data']['addition'] = 56 if message['data']
      callback.call(message)
    }
    
    publish :A, '/channels/b', 'message_for' => 'B'
    check_inbox(
        :A => {},
        :B => {
          '/channels/b' => ['message_for' => 'B', 'addition' => 56]
        }
    )
  end
  
  scenario "Server blocks outgoing message" do
    server 8000
    http_client :A, []
    http_client :B, ['/channels/b']
    
    extend_server :outgoing, lambda { |message, callback|
      if !message['data'] or message['data']['deliver'] == 'yes'
        callback.call(message)
      else
        callback.call(nil)
      end
    }
    
    publish :A, '/channels/b', [{'deliver' => 'no'}, {'deliver' => 'yes'}]
    
    check_inbox(
        :A => {},
        :B => {
          '/channels/b' => ['deliver' => 'yes']
        }
    )
  end
  
  scenario "Two HTTP clients, no messages delivered" do
    server 8000
    http_client :A, ['/channels/a']
    http_client :B, []
    publish :B, '/channels/b', 'hello' => 'world'
    check_inbox(
      :A => {},
      :B => {}
    )
  end
  
  scenario "Two HTTP clients, single subscription" do
    server 8000
    http_client :A, ['/channels/a']
    http_client :B, []
    publish :B, '/channels/a', 'hello' => 'world'
    check_inbox(
        :A => {
          '/channels/a' => ['hello' => 'world']
        },
        :B => {}
    )
  end

  scenario "Two HTTP clients, multiple subscriptions" do
    server 8000
    http_client :A, ['/channels/a', '/channels/*']
    http_client :B, []
    publish :B, '/channels/a', 'hello' => 'world'
    check_inbox(
        :A => {
          '/channels/a' => ['hello' => 'world'],
          '/channels/*' => ['hello' => 'world']
        },
        :B => {}
    )
  end

  scenario "Three HTTP clients, single receiver" do
    server 8000
    http_client :A, ['/channels/a']
    http_client :B, []
    http_client :C, ['/channels/c']
    publish :B, '/channels/a', 'chunky' => 'bacon'
    check_inbox(
        :A => {
          '/channels/a' => ['chunky' => 'bacon']
        },
        :B => {},
        :C => {}
    )
  end

  scenario "Three HTTP clients, multiple receivers" do
    server 8000
    http_client :A, ['/channels/shared']
    http_client :B, []
    http_client :C, ['/channels/shared']
    publish :B, '/channels/shared', 'chunky' => 'bacon'
    check_inbox(
        :A => {
          '/channels/shared' => ['chunky' => 'bacon']
        },
        :B => {},
        :C => {
          '/channels/shared' => ['chunky' => 'bacon']
        }
    )
  end

  scenario "Two HTTP clients, single wildcard on receiver" do
    server 8000
    http_client :A, ['/channels/*']
    http_client :B, []
    publish :B, '/channels/anything', 'msg' => 'hey'
    check_inbox(
        :A => {
          '/channels/*' => ['msg' => 'hey']
        },
        :B => {}
    )
  end

  scenario "Two HTTP clients, single wildcard on sender" do
    server 8000
    http_client :A, ['/channels/name', '/channels/hello', '/channels/nested/hello']
    http_client :B, []
    publish :B, '/channels/*', 'msg' => 'hey'
    check_inbox(
        :A => {
          '/channels/name' => ['msg' => 'hey'],
          '/channels/hello' => ['msg' => 'hey']
        },
        :B => {}
    )
  end

  scenario "Two HTTP clients, single wildcard on both" do
    server 8000
    http_client :A, ['/channels/*']
    http_client :B, []
    publish :B, '/channels/*', 'msg' => 'hey'
    check_inbox(
        :A => {
          '/channels/*' => ['msg' => 'hey']
        },
        :B => {}
    )
  end

  scenario "Two local clients, double wildcard on sender" do
    server 8000
    local_client :A, ['/channels/name', '/channels/hello', '/channels/nested/hello']
    local_client :B, []
    publish :B, '/channels/**', 'msg' => 'hey'
    check_inbox(
        :A => {
          '/channels/name' => ['msg' => 'hey'],
          '/channels/hello' => ['msg' => 'hey'],
          '/channels/nested/hello' => ['msg' => 'hey']
        },
        :B => {}
    )
  end

  scenario "Two local clients, one HTTP, double wildcard on sender and one subscription" do
    server 8000
    local_client :A, ['/channels/hello', '/channels/nested/hello']
    local_client :B, []
    http_client :C, ['/channels/name', '/channels/foo/**']
    publish :B, '/channels/**', 'msg' => 'hey'
    check_inbox(
        :A => {
          '/channels/hello' => ['msg' => 'hey'],
          '/channels/nested/hello' => ['msg' => 'hey']
        },
        :B => {},
        :C => {
          '/channels/name' => ['msg' => 'hey'],
          '/channels/foo/**' => ['msg' => 'hey'],
          '/channels/name' => ['msg' => 'hey'],
        }
    )
  end
end

